Added barrier sync functionality.

andrewbenton opened the pull request:
Public API is exposed in the Barrier class.  This allows test scenarios to synchronize between themselves, and includes support for tests not making the barrier, timeouts, and distinct code paths.  Code operating under this barrier sync implementation may not, and likely will not, execute properly when mixing the barrier sync with a serialized test, however it will still run.

andrewbenton committed:
Added barrier sync functionality for tests.

Adds a barrier sync function through Barrier.sync().  All other API
changes should be internal for future flexibility.  Internally
introduced classes include FlexibleBarrier, which is used for the
barrier sync, and StaticBarrier, which is used for the sync prior to the
scenario beginning.  The ZucchiniInvocationHandler, ZucchiniRuntime, and
ZucchiniRuntimeOptions were also introduced and provide internal
functionality that wraps around Cucummber and exposes the same API.

Logback was also changed so that certain classes, like exceptions, would
flag colored output.

andrewbenton committed:
Changed barrier sync wait interpretation.

Previously, barrier sync wait times less than 1 resulted in a
timeout-less wait.  This has been changed so that if the timeout value
is set to 0, it will pass through and skip the sync, and if it is
less than 0, it will wait for an indefinite amount of time.

andrewbenton committed:
Altered internal barriers to work with OpenJVM.

Previous code using CountDownLatch seemed to have issues with OpenJVM,
though it worked on OracleJVM.  This switches the internal
implementations to use the Phaser instead, which isn't a perfect fit,
but it doesn't cause the same issue on OpenJVM.  Needs to be tested on
OracleJVM.

andrewbenton committed:
Added documentation to recent changes.

Added documentation to all of the new classes and methods, as well as
some of the older methods that were still recently added, but lack
documentation.  At a later point, I will go through and add more
documentation to the classes that wrap Cucumber, because cucumber
provides no documentation for many of their public methods.

andrewbenton committed:
Converted the slf4j loggers to be named 'LOGGER'.

This was a change for convention, as all Loggers should be private
static final, and so should be named as constants.

andrewbenton committed:
Enabled serialized runs with barrier sync.

Serialized runs will not necessarily pass, as barrier syncs do not cause
the thread to yield and switch to another thread for concurrent
execution, but it will prevent the serialized run from permanently
halting.

andrewbenton committed:
Formatting fixes.

andrewbenton committed:
Race prevention with double-checked locking.

The addition of failed contexts had the potential to be unstable before.
This should help to make it more stable by ensuring that the actions
that are bundled with adding the failed context to the set are executed
only once.  I may need to break more rules at look at thread
pause/resume actions in the future.

andrewbenton committed:
Removed the gmaven plugin and requirement.

Since groovy is no longer needed, the gmaven plugin isn't needed.

andrewbenton committed:
Adds more protection for zucchini's execution.

Prevents threads from being killed when they are unable to catch the
ThreadDeath call.  This should help to prevent pre-kill or post-kill
(when the thread is running but is either past, or not yet in, a
scenario), though it doesn't help with potential blocking, as threads in
the pre state may not recognize that they have been killed, and will
then try to proceed.

andrewbenton committed:
Merge branch 'feature/barrier_sync_v2' of github.com:andrewbenton/zucchini into feature/barrier_sync_v2

andrewbenton committed:
Safer handling of failedContexts.

Switched the AbstractZucchiniTest.failedContexts to use a
ConcurrentHashMap to extract a Set instead of the HashSet.  Also added a
check to the FlexibleBarrier.await to check if the context has already
failed, and it if has, then the thread will stop itself.

andrewbenton committed:
Applied patch from @TrentonTrees for clean imports

andrewbenton committed:
Moved verbose messages to Trace level.

andrewbenton committed:
Safer synchronization for AZT and TestContext

andrewbenton committed:
Various safety, correctness, and formatting changes.

andrewbenton committed:
Merge pull request #3 from andrewbenton/feature/barrier_sync_v2

Feature/barrier sync v2

andrewbenton committed:
Merge branch 'develop' of https://github.com/Comcast/zucchini into Comcast-develop

andrewbenton committed:
Altered force-fail run state to default off.

This is done for the sake of build systems.  This can be re-enabled for
testing.

andrewbenton committed:
Merge branch 'Comcast-develop' into develop

andrewbenton committed:
Removed getOrDefault, as it caused problems.

andrewbenton committed:
Removed explicit env variable from travis.yml

andrewbenton committed:
Removed reliance on MojoExecutionException.

This reliance was causing issues on some systems.  This should fix the
problem.

dimaj commented on pom.xml:
```diff
@@ -81,26 +80,6 @@
             </plugin>
 
             <plugin>
-                <groupId>org.codehaus.gmaven</groupId>
```
just wondering... what were these plugins used for and why did you remove them?

andrewbenton replied:
Groovy integration for apache maven: https://groovy.github.io/gmaven/

dimaj commented on src/main/java/com/comcast/zucchini/AbstractZucchiniTest.java:
```diff
@@ -12,12 +17,13 @@
 import com.google.gson.JsonElement;
 import com.google.gson.JsonParser;
 
+import cucumber.runtime.CucumberException;
```
this exception is not being used, please remove

andrewbenton replied:
Gladly :)

dimaj commented on src/main/java/com/comcast/zucchini/FlexibleBarrier.java:
```diff
@@ -0,0 +1,234 @@
+                            this.azt.failedContexts.add(tc);
+                            this.dec();
+                            if(tc.canKill) {
+                                tc.getThread().stop();
```
Thead.stop() has been deprecated...
http://stackoverflow.com/questions/10961714/how-to-properly-stop-the-thread-in-java

andrewbenton replied:
It's unfortunately necessary, given that we are trying to prevent bad code from causing Zucchini to block on the barrier sync.  The Thread.interrupt() is only useful if the code within the thread is checking for interruptions or is performing IO operations.  

The Thread.stop() method will allow us to break from bad code such as 'while(true) {}'.  If I remove this, then we can't enforce the barrier timeout. @trentontrees

trentontrees replied:
I'm fine with the approach that you have supplied here using Thread.stop().

dimaj commented on src/main/java/com/comcast/zucchini/AbstractZucchiniTest.java:
```diff
@@ -31,18 +37,34 @@
     private static Boolean hooked = false;
 
+    /* store the list of contexts here */
+    List<TestContext> contexts;
```
should these vars be private with a getter?

andrewbenton replied:
It's used at package-level access.  For now, it's not a part of the API, and people inheriting from AZT aren't able to access it anyway.  Providing a getter would be an unnecessary extra step.

andrewbenton replied:
I'll take a look to check and be sure that I'm using it though.  I'm pretty sure that I use it in sizing the barriers auto-magically.

dimaj commented on src/main/java/com/comcast/zucchini/AbstractZucchiniTest.java:
```diff
@@ -86,21 +140,25 @@ public void runParallel(List<TestContext> contexts) {
-            catch(Throwable e) {
-                logger.error(t.toString());
+            catch(InterruptedException e) {
+                LOGGER.error("ERROR on Thread.join(): {}", e);
```
if there was an exception while joining threads, should tests fail?

andrewbenton replied:
That might be a good idea, thought the fail should wait until they have all joined or failed, if we choose to go that route.  We might need to add the MojoExecutionException back in as the 'throw' if we want it run later tests.  

On a side note, I don't think that the InterruptedException can make it out to this point unless the interrupted flag is set and is then allowed to propagate.  If it is raised in on the of the step definitions, it won't make it back up to this level.  I'll have to check the docs again to be sure.

andrewbenton committed:
Minor changes in AZT for thread join.

Added check for thread join and better error message for when any
contexts encountered errors.  As a side note, the imports were
rearranged to be more correct.

dimaj commented on src/main/java/com/comcast/zucchini/FlexibleBarrier.java:
```diff
@@ -0,0 +1,234 @@
+    /**
+     * A shortcut to get the {@see TestContext} name or provide "<NULL>" it its place.
+     */
+    private static String name() {
```
Factor this out to a utility class as it is used / defined in multiple places. You can use `import static` in your other classes to save on typing.

dimaj commented on src/main/java/com/comcast/zucchini/ZucchiniRuntime.java:
```diff
@@ -0,0 +1,276 @@
+     * {@inheritDoc}
+     */
+    @Override
+    public void printSummary() {
```
remove all methods that just call `super.methodName`

dimaj commented on src/main/java/com/comcast/zucchini/ZucchiniShutdownHook.java:
```diff
@@ -37,22 +39,29 @@ public void run() {
+        finally {
+            if(writer != null) {
+                try {
+                writer.close();
```
please indent

andrewbenton committed:
Factored out name() static. Minor fixes for merge.

andrewbenton committed:
Merge branch 'develop' of github.com:Comcast/zucchini into develop
